From e9fc79d8935409f8e48a21f8a6e5d4cf50aa65b6 Mon Sep 17 00:00:00 2001
From: "zhenjun.zhang" <zhenjun.zhang@spreadtrum.com>
Date: Mon, 5 Feb 2018 10:57:29 +0800
Subject: [PATCH 144/296] Systemserver crash issue.

The LegacyGlobalAction's dialog may be create in two thread, since
view's UI operation can only do in one thread, it may occur systemserver
crash during monkey.

This issue may reproduce when SystemUI is crash or killing when ANR.

Fix: create the dialog useing the handler thread only.
---
 services/core/java/com/android/server/policy/GlobalActions.java | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/services/core/java/com/android/server/policy/GlobalActions.java b/services/core/java/com/android/server/policy/GlobalActions.java
index 342ec4b79fe..942d049c0ea 100644
--- a/services/core/java/com/android/server/policy/GlobalActions.java
+++ b/services/core/java/com/android/server/policy/GlobalActions.java
@@ -90,8 +90,10 @@ class GlobalActions implements GlobalActionsListener {
         mStatusBarConnected = connected;
         if (mShowing && !mStatusBarConnected) {
             // Status bar died but we need to be showing global actions still, show the legacy.
-            ensureLegacyCreated();
-            mLegacyGlobalActions.showDialog(mKeyguardShowing, mDeviceProvisioned);
+            mHandler.post(() -> {
+                ensureLegacyCreated();
+                mLegacyGlobalActions.showDialog(mKeyguardShowing, mDeviceProvisioned);
+            });
         }
     }
 
-- 
2.11.0

